core/consensus/babe/src/lib.rs

```diff
@@ -95,8 +203,845 @@ impl<Hash> CompatibleDigestItem for generic::DigestItem<Hash, Public, Secret> {
+//
+// FIXME #1018 needs misbehavior types
+#[forbid(warnings)]
+fn check_header<B: Block + Sized>(
+	slot_now: u64,
+	mut header: B::Header,
+	hash: B::Hash,
+	authorities: &[Public],
+) -> Result<CheckedHeader<B::Header, DigestItemFor<B>>, String>
+	where DigestItemFor<B>: CompatibleDigestItem,
+{
+	trace!("Checking header");
+	let digest_item = match header.digest_mut().pop() {
+		Some(x) => x,
+		None => return Err(format!("Header {:?} is unsealed", hash)),
+	};
+
+	let BabeSeal { slot_num, signature: LocalizedSignature { signer, signature }, proof, vrf_output } = digest_item.as_babe_seal().ok_or_else(|| {
+		error!(target: "babe", "Header {:?} is unsealed", hash);
+		format!("Header {:?} is unsealed", hash)
+	})?;
+	// FIXME!!
+	let threshold = u64::MAX;
+	if slot_num > slot_now {
+		header.digest_mut().push(digest_item);
+		Ok(CheckedHeader::Deferred(header, slot_num))
+	} else if !authorities.contains(&signer) {
+		error!(target: "babe", "Slot Author not found");
+		Err("Slot Author not found".to_string())
+	} else {
+		let pre_hash = header.hash();
+		let to_sign = (slot_num, pre_hash, proof.to_bytes()).encode();
+
+		if sr25519::Pair::verify(&signature, &to_sign[..], &signer) {
+			let (inout, _batchable_proof) = {
+				let transcript = make_transcript(
+					Default::default(),
+					slot_num,
+					Default::default(),
+					0,
+				);
+				schnorrkel::PublicKey::from_bytes(signer.as_slice()).and_then(|p| p.vrf_verify(transcript, &vrf_output, &proof)).map_err(|s| {
+					warn!(target: "babe", "VRF verification failed: {:?}", s);
+					format!("VRF verification failed")
+				})?
+			};
+			let r: u64 = inout.make_chacharng(b"substrate-babe-vrf").gen();
+			if r < threshold {
+				Ok(CheckedHeader::Checked(header, digest_item))
+			} else {
+				error!(target: "babe", "VRF verification failed: threshold {} but VRF produced {}", r, threshold);
+				Err(format!("Validator {:?} made seal when it wasn’t its turn", signer))
+			}
+		} else {
+			error!(target: "babe", "Bad signature on {:?}", hash);
+			Err(format!("Bad signature on {:?}", hash))
+		}
+	}
+}
+
+/// Extra verification for Babe blocks.
+pub trait ExtraVerification<B: Block>: Send + Sync {
+	/// Future that resolves when the block is verified or fails with error if not.
+	type Verified: IntoFuture<Item=(),Error=String>;
+
+	/// Do additional verification for this block.
+	fn verify(
+		&self,
+		header: &B::Header,
+		body: Option<&[B::Extrinsic]>,
+	) -> Self::Verified;
+}
+
+/// A verifier for Babe blocks.
+pub struct BabeVerifier<C, E> {
+	client: Arc<C>,
+	extra: E,
+	inherent_data_providers: inherents::InherentDataProviders,
+}
+
+impl<C, E> BabeVerifier<C, E> {
+	fn check_inherents<B: Block>(
+		&self,
+		block: B,
+		block_id: BlockId<B>,
+		inherent_data: InherentData,
+		timestamp_now: u64,
+	) -> Result<(), String>
+		where C: ProvideRuntimeApi, C::Api: BlockBuilderApi<B>
+	{
+		const MAX_TIMESTAMP_DRIFT_SECS: u64 = 60;
```

Comment:
Given that BABE slots can't be unilaterally associated with a time (due to relative timing), this whole function may be unnecessary.

Reply:
or rather the whole `ValidAtTimestamp` branch